pkg/spiffs/fs/spiffs_fs.c

```diff
@@ -0,0 +1,429 @@
+        s_flags |= SPIFFS_O_TRUNC;
+    }
+    if (flags & O_CREAT) {
+        s_flags |= SPIFFS_O_CREAT;
+    }
+    if (flags & O_WRONLY) {
+        s_flags |= SPIFFS_O_WRONLY;
+    }
+    if (flags & O_RDWR) {
+        s_flags |= SPIFFS_O_RDWR;
+    }
+#ifdef __O_DIRECT
+    if (flags & __O_DIRECT) {
+        s_flags |= SPIFFS_O_DIRECT;
+    }
+#endif
+    if (flags & O_EXCL) {
+        s_flags |= SPIFFS_O_EXCL;
+    }
+
+    DEBUG("spiffs: open: %s (abs_path: %s), flags: 0x%x, mode: %d\n", name, abs_path, s_flags, mode);
+
+    s32_t ret = SPIFFS_open(&fs_desc->fs, name, s_flags, mode);
+    if (ret >= 0) {
+        filp->private_data.value = ret;
+        return ret;
+    }
+    else {
+        return spiffs_err_to_errno(ret);
+    }
+}
+
+static int _close(vfs_file_t *filp)
+{
+    spiffs_desc_t *fs_desc = filp->mp->private_data;
+
+    return spiffs_err_to_errno(SPIFFS_close(&fs_desc->fs, filp->private_data.value));
+}
+
+static ssize_t _write(vfs_file_t *filp, const void *src, size_t nbytes)
+{
+    spiffs_desc_t *fs_desc = filp->mp->private_data;
+
+    return spiffs_err_to_errno(SPIFFS_write(&fs_desc->fs, filp->private_data.value, src, nbytes));
+}
+
+static ssize_t _read(vfs_file_t *filp, void *dest, size_t nbytes)
+{
+    spiffs_desc_t *fs_desc = filp->mp->private_data;
+
+    return spiffs_err_to_errno(SPIFFS_read(&fs_desc->fs, filp->private_data.value, dest, nbytes));
+}
+
+static off_t _lseek(vfs_file_t *filp, off_t off, int whence)
+{
+    spiffs_desc_t *fs_desc = filp->mp->private_data;
+
+    int s_whence = 0;
+    if (whence == SEEK_SET) {
+        s_whence = SPIFFS_SEEK_SET;
+    }
+    else if (whence == SEEK_CUR) {
+        s_whence = SPIFFS_SEEK_CUR;
+    }
+    else if (whence == SEEK_END) {
+        s_whence = SPIFFS_SEEK_END;
+    }
+
+    return spiffs_err_to_errno(SPIFFS_lseek(&fs_desc->fs, filp->private_data.value, off, s_whence));
+}
+
+static int _fstat(vfs_file_t *filp, struct stat *buf)
+{
+    spiffs_desc_t *fs_desc = filp->mp->private_data;
+    spiffs_stat stat;
+    s32_t ret;
+
+    memset(buf, 0, sizeof(*buf));
+
+    ret = SPIFFS_fstat(&fs_desc->fs, filp->private_data.value, &stat);
+
+    if (ret < 0) {
+        return ret;
+    }
+    //stat.name;
+    buf->st_ino = stat.obj_id;
+    //stat.pix;
+    buf->st_size = stat.size;
+    //stat.type;
+    buf->st_mode = S_IFREG;
+
+    return spiffs_err_to_errno(ret);
+}
+
+static int _opendir(vfs_DIR *dirp, const char *dirname, const char *abs_path)
+{
+    spiffs_desc_t *fs_desc = dirp->mp->private_data;
+    spiffs_DIR d;
+    (void) abs_path;
+
+    if (VFS_DIR_BUFFER_SIZE < sizeof(spiffs_DIR)) {
+        return -EOVERFLOW;
+    }
+
+    SPIFFS_opendir(&fs_desc->fs, dirname, &d);
+    memcpy(dirp->private_data.buffer, &d, sizeof(d));
+
+    return 0;
+}
+
+static int _readdir(vfs_DIR *dirp, vfs_dirent_t *entry)
+{
+    spiffs_DIR d;
+    struct spiffs_dirent e;
+    struct spiffs_dirent *ret;
+
+    memcpy(&d, dirp->private_data.buffer, sizeof(d));
```

Comment:
What is the purpose of the memcpy back and forth for the call to `SPIFFS_readdir`?